Red tech

redtechnologies opened the pull request:
in this request we added methods to the SAS admin interface to:
- inject Esc Zone
- trigger  Esc Zone
- reset trigger  of Esc Zone

and we added two tests that use those methods in grant test case :

- GRA.2
- GRA.3

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -176,7 +176,15 @@ def _GetSuccessResponse(self):
 
-
+  def InjectEscZone(self, request, ssl_cert=None, ssl_key=None)
+    zoneData = req['zoneDate']
```
This should be:

```
zone_data = request['zoneData']
```

1) We've been using snake_case (as opposed to camelCase) variable names in the rest of the code.
2) "req" is undefined so we use "request"
3) Typo: "zoneDate"

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -176,7 +176,15 @@ def _GetSuccessResponse(self):
+  def InjectEscZone(self, request, ssl_cert=None, ssl_key=None)
+    zoneData = req['zoneDate']
+    if(zoneData['usage'] == 'EXCLUSION_ZONE'):
+      idAttributes = zoneData['id'].split('/')
```
Why can't we just return zoneData['id'] directly? Fake SAS isn't supposed to be very smart, it just needs to return things which are syntactically correct.

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -176,7 +176,15 @@ def _GetSuccessResponse(self):
+      zoneId = ''; 
+     for zoneIdPart in idAttributes[4 : len(idAttributes)]:
+       zoneId+=zoneIdPart
+     return {'zone_id' : zoneId}
```
What happens is usage isn't 'EXCLUSION_ZONE'? zoneId won't be defined and this will cause an error. We probably want to return an HTTP error in this case (or just ignore the problem and return a string anyway).

kate-harrison replied:
I apologize: my doc is misleading in this regard. I used "zone_id" in the doc but intended to use "zoneId" in the code since that is consistent with the style used in other SAS APIs. Can you please update this accordingly?

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -197,8 +205,12 @@ def do_POST(self):
+    elif self.path == '/admin/injectdata/esc_zone':
+      response = FakeSas().InjectEscZone(request)
+    elif self.path == '/admin/trigger/esc_detection/':
+	  response = {'trigger_id': 'fake_trigger_id'}
```
Please make indentation consistent with the nearby code.

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -178,6 +177,39 @@ def InjectFccId(self, request):
+  def InjectEscZone(self, request):
+    """SAS admin interface to inject EscZone information into SAS under test.
+    Args:
+      request: ey-value pair where the key is
```
Typo: "key-value"

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -178,6 +177,39 @@ def InjectFccId(self, request):
+    Args:
+      request: ey-value pair where the key is
+        "zone" and the value is  he information of the zone of type ZoneData as specified in specified in SAS-SAS TS:
+		id	string with format zone/$CREATOR/$ZONE_ID
```
I think it's fine to omit this additional information. Anyone reading this code will have easy access to the SAS-SAS TS so including it won't save them much trouble. Having short comments will make it easier for someone to read through the file and find what they're looking for. It also means that we don't have to update the comment if the TS changes.

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -189,4 +221,26 @@ def PreloadRegistrationData(self, request):
+  def TriggerEscZone(self, request):
+	"""SAS admin interface to Trigger Esc Zone in SAS under test.
+	Args:
+		  request: A dictionary with a single key-value pair where one key is
```
Reduce indentation? I think it will make the rest of the comment easier to read (and it will be consistent with nearby code).

kate-harrison commented on src/harness/testcases/testdata/esc_zone_contains_device_a.json:
```diff
@@ -0,0 +1,10 @@
+{"id" :"zone/exclusion_zone/ntia/2017_04_10/esc_zone_contains_deviceA_harness_test",
+			 "creator" : "SAS_Administrator_Test",
```
Consider reducing indentation for readability.

kate-harrison commented on src/harness/testcases/grant_testcase.py:
```diff
@@ -133,6 +133,102 @@ def test_10_7_4_1_3_1_2_1(self):
+    # Create and trigger the ESC Zone
+    esc_zone_not_contain_device_a = json.load(
+        open(os.path.join('testcases', 'testdata', 'esc_zone_not_contain_device_a.json')))
+    zoneMessage = {'zone':esc_zone_not_contain_device_a}
```
Please use `request` or `zone_request` to remain consistent with our variable naming convention.

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -177,14 +177,15 @@ def _GetMissingParamResponse(self):
+      for zone_id_part in id_attributes[4 : len(id_attributes)]:
+        zone_id += zone_id_part
+      return {'zone_id' : zone_id}
+	return ''
```
I don't think this code can run successfully with this indentation.

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -177,14 +177,15 @@ def _GetMissingParamResponse(self):
-     return {'zone_id' : zoneId}
-		
+    zone_data = request['zoneData']
+    if(zone_data['usage'] == 'EXCLUSION_ZONE'):
```
Why not just return the zone ID directly? I don't think it's necessary to parse and then recombine the ID.

kate-harrison replied:
This actually raises a good point: ZoneData isn't originally meant to be used for ESC so it's not clear what the value of "usage" should be in this case. Maybe allowing the field to be empty is better than using something that may be misleading?

Whatever we decide, we should make sure we write it down in the API doc.

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -179,31 +179,10 @@ def InjectFccId(self, request):
     Args:
-      request: ey-value pair where the key is
+      request: key-value pair where the key is
         "zone" and the value is  he information of the zone of type ZoneData as specified in specified in SAS-SAS TS:
```
Suggestion:

```
"zoneData": (ZoneData object, as defined in the SAS-SAS TS) the zone to be injected
```

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -177,14 +177,15 @@ def _GetMissingParamResponse(self):
+      zone_id = ''; 
+      for zone_id_part in id_attributes[4 : len(id_attributes)]:
+        zone_id += zone_id_part
+      return {'zone_id' : zone_id}
```
In the dictionary, it should be "zoneId" for consistency with other messages.

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -226,10 +205,11 @@ def PreloadRegistrationData(self, request):
-			"zone_id" a value of string type.
-			the other key is "frequency_range" with a value of type FrequencyRange object as specified in SAS-CBSD TS
+      request: A dictionary with a single key-value pair where one key is
+    	"zone_id" a value of string type.
```
Here we should have "zoneId" and "frequencyRange".

kate-harrison commented on src/harness/testcases/testdata/esc_zone_contains_device_a.json:
```diff
@@ -1,10 +1,18 @@
-						 [37.125056, -122.284112],[37.125056, -121.884112],[37.525056, -121.884112]]]
-			}
+{
+ "id" :"zone/exclusion_zone/ntia/2017_04_10/esc_zone_contains_deviceA_harness_test",
```
ESC zones aren't created by the NTIA so I'm not sure this is a good format for the ID. Maybe its best to say that this field will not be filled in by the test harness (instead the SAS under test will generate the ID) since there's no standard ID format for ESCs?

krlinus commented on src/harness/fake_sas.py:
```diff
@@ -176,7 +176,16 @@ def _GetSuccessResponse(self):
+    zone_data = request['zoneData']
+    if(zone_data['usage'] == 'EXCLUSION_ZONE'):
+      id_attributes = zone_data['id'].split('/')
+      zone_id = ''; 
```
semi-colon not required

krlinus commented on src/harness/fake_sas.py:
```diff
@@ -176,7 +176,16 @@ def _GetSuccessResponse(self):
+      zone_id = ''; 
+      for zone_id_part in id_attributes[4 : len(id_attributes)]:
+        zone_id += zone_id_part
+      return {'zone_id' : zone_id}
```
Is the intent here is to re-create the zoneId from id field, by removing the '/'es starting from the 4th slash?

kate-harrison commented:
I see some new commits but it looks like this isn't ready for review yet. Please let us know explicitly when you'd like another round of review. Thanks!

redtechnologies commented:
We fixed conflicts with the master, added Trigger, Inject and reset Esc zone to sas_interface.py.
In fake_sas.py we changed trigger_id to triggerId to be coherent with the other docs.
Thanks for the quick reviews and feedbacks.

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -209,6 +209,10 @@ def do_POST(self):
                        '/admin/injectdata/cluster_list',
                        '/admin/injectdata/pal_database_record'):
       response = ''
+
```
Consider removing this extra line for readability.

kate-harrison commented on src/harness/fake_sas.py:
```diff
@@ -209,6 +209,10 @@ def do_POST(self):
                        '/admin/injectdata/cluster_list',
                        '/admin/injectdata/pal_database_record'):
```
I think you need to add "esc_reset" to this list and "esc_zone" in its own 'elif' below. I thought they were in your earlier code; did you take them out? It looks like InjectEscZone() also disappeared. Am I missing something?

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -200,9 +200,21 @@ def BlacklistByFccIdAndSerialNumber(self, request):
+     
+    Args:
+      request: key-value pair where the key is
+        "zone" and the value is  he information of the zone of type ZoneData as specified in specified in SAS-SAS TS:
```
Can we use "zoneData" to be consistent with other methods?

Let's try:
... "zoneData" and the value is ZoneData structure as defined in the SAS-SAS TS (WINNF-16-S-0096).

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -200,9 +200,21 @@ def BlacklistByFccIdAndSerialNumber(self, request):
+    Args:
+      request: key-value pair where the key is
+        "zone" and the value is  he information of the zone of type ZoneData as specified in specified in SAS-SAS TS:
+	Returns: string zone_id
```
"zoneId" (camelCase instead of snake_case)

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -252,17 +264,39 @@ def InjectPalDatabaseRecord(self, request):
+
+      Args:
+       request: A dictionary with a single key-value pair where one key is
+         "zone_id" a value of string type.
```
"zoneId"

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -252,17 +264,39 @@ def InjectPalDatabaseRecord(self, request):
+      Args:
+       request: A dictionary with a single key-value pair where one key is
+         "zone_id" a value of string type.
+         the other key is "frequency_range" with a value of type FrequencyRange object as specified in SAS-CBSD TS
```
"frequencyRange"

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -252,17 +264,39 @@ def InjectPalDatabaseRecord(self, request):
+       request: A dictionary with a single key-value pair where one key is
+         "zone_id" a value of string type.
+         the other key is "frequency_range" with a value of type FrequencyRange object as specified in SAS-CBSD TS
+       Returns: string trigger_id
```
"triggerId"

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -252,17 +264,39 @@ def InjectPalDatabaseRecord(self, request):
 
   @abc.abstractmethod
+  def TriggerEscZone(self, request):
+      """SAS admin interface to Trigger Esc Zone in SAS under test.
```
Do you want to mention the behavior in this comment?

kate-harrison commented on src/harness/sas_interface.py:
```diff
@@ -252,17 +264,39 @@ def InjectPalDatabaseRecord(self, request):
+
+  @abc.abstractmethod
+  def ResetEscZone(self, request):
+      """SAS admin interface to reset the Trigger of Esc Zone in SAS under test.
```
Similar comments here as on TriggerEscZone().

kate-harrison commented on src/harness/testcases/grant_testcase.py:
```diff
@@ -133,6 +133,99 @@ def test_10_7_4_1_3_1_2_1(self):
+        open(os.path.join('testcases', 'testdata', 'esc_zone_not_contain_device_a.json')))
+    zone_request= {'zone':esc_zone_not_contain_device_a}
+    zone_response = self._sas_admin.InjectEscZone(zone_request)
+    trigger_esc_zone_request = {'zone_id': zone_response['zone_id'],
```
"zoneId"

kate-harrison commented on src/harness/testcases/grant_testcase.py:
```diff
@@ -133,6 +133,99 @@ def test_10_7_4_1_3_1_2_1(self):
+    zone_request= {'zone':esc_zone_not_contain_device_a}
+    zone_response = self._sas_admin.InjectEscZone(zone_request)
+    trigger_esc_zone_request = {'zone_id': zone_response['zone_id'],
+                                    'frequency_range': {
```
"frequencyRange" (please check all other JSON field names in this file)

kate-harrison commented on src/harness/testcases/testdata/esc_zone_contains_device_a.json:
```diff
@@ -0,0 +1,18 @@
+{
+ "id" :"zone/exclusion_zone/ntia/2017_04_10/esc_zone_contains_deviceA_harness_test",
```
I don't think the ID can have this format since the zone is not created by the NTIA. I'm not sure if there's any guidance on the ID format, so we can probably use a format like:

"zone/esc_zone/sas_admin/zone_id"

kate-harrison commented on src/harness/testcases/testdata/esc_zone_contains_device_a.json:
```diff
@@ -0,0 +1,18 @@
+{
+ "id" :"zone/exclusion_zone/ntia/2017_04_10/esc_zone_contains_deviceA_harness_test",
+ "creator" : "SAS_Administrator_Test",
+ "usage" : "EXCLUSION_ZONE",
```
I think it's best if we leave "usage" empty.

kate-harrison commented on src/harness/testcases/testdata/esc_zone_contains_device_a.json:
```diff
@@ -0,0 +1,18 @@
+{
+ "id" : "zone/exclusion_zone/ntia/2017_04_10/esc_zone_contains_deviceA_harness_test",
```
As mentioned earlier, this ID format is not appropriate for an ESC.

kate-harrison commented on src/harness/testcases/testdata/esc_zone_contains_device_a.json:
```diff
@@ -0,0 +1,18 @@
+{
+ "id" : "zone/exclusion_zone/ntia/2017_04_10/esc_zone_contains_deviceA_harness_test",
+ "creator" : "SAS_Administrator_Test",
+ "usage" : "EXCLUSION_ZONE",
```
As mentioned earlier, please remove this field.

kate-harrison commented on src/harness/testcases/testdata/esc_zone_contains_device_a.json:
```diff
@@ -0,0 +1,18 @@
+  "coordinates":
+  [
+   [
+    [37.525056, -122.284112],
```
FYI device A has moved to Kansas to make other test cases easier. It's probably best to define an ESC zone along the coast and then move device A inside or nearby as needed.

janetliuboston commented:
We're having a hard time following all the changes on this branch, knowing when they are done, and are not able to approve just a part of the PR. Would you be able to separate out each test case and framework into a separate PR? This way they will be easier to comment and approve on contained changes, and know when a change to a comment has been completed and is ready to be reviewed again. Thank you.
